sys/linux/io_uring.txt

```diff
@@ -3,56 +3,337 @@
+type io_uring_sqe_provide_buffers io_uring_sqe_t[IORING_OP_PROVIDE_BUFFERS, const[0, int16], int32, io_uring_bid[int64], buffer[in], int32, const[0, int32], sqe_user_data_not_openat, buf_group_personality_misc]
+type io_uring_sqe_remove_buffers io_uring_sqe_t[IORING_OP_PROVIDE_BUFFERS, const[0, int16], int32, const[0, int64], const[0, int64], const[0, int32], const[0, int32], sqe_user_data_not_openat, buf_group_personality_misc]
+type io_uring_sqe_tee io_uring_sqe_t[IORING_OP_TEE, const[0, int16], fd_or_fixed_fd_index, const[0, int64], const[0, int64], int32, flags[splice_flags, int32], sqe_user_data_not_openat, io_uring_sqe_tee_misc]
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+# 
+# Flags, enumerations, and misc fields of sqe ops
+#
+
+iosqe_flags = IOSQE_IO_DRAIN_BIT, IOSQE_IO_LINK_BIT, IOSQE_IO_HARDLINK_BIT, IOSQE_ASYNC_BIT, IOSQE_BUFFER_SELECT_BIT, IOSQE_FIXED_FILE_BIT
+
+fd_or_fixed_fd_index [
+	fd		fd
+# Use the registered files (io_uring_register$IORING_REGISTER_FILES) when IOSQE_FIXED_FILE_BIT is set in sqe.
+# To ease collisions, limit the indices.
+	fd_index	int32[0:10]
+]
+
+# 0 for normal file integrity sync, IORING_FSYNC_DATASYNC to provide data sync only semantics
+io_uring_fsync_flags = 0, IORING_FSYNC_DATASYNC
+
+# 0 for relative, IORING_TIMEOUT_ABS for absolute timeout value
+io_uring_timeout_flags = 0, IORING_TIMEOUT_ABS
+
+# The timeout condition is met when either the specific timeout expries, or the
+# specified number of events have completed. If not set, defaults to 1. Use a
+# limited range to allow utilization of this value to meet timeout condition besides
+# the timeout expiration.
+type io_uring_timeout_completion_event_count int64[0:10]
+
+# An already issued request can be attempted to be cancelled using ASYNC_CANCEL
+# operation. This operation identifies the operations using what's passed as
+# with user_data in their sqe. To ease collisions of ASYNC_CANCEL operation with
+# already submitted ones, use a limited range of values for user_data field.
+# Among all operations that can be achieved by submitting to the io_uring, only
+# openat and openat2 returns a useful resource (fd) that we can use for other
+# systemcalls. The resulting fds are returned within io_uring_cqe.res. The only way
+# to identify cqes for those operations is to keep track of their user data. Thus,
+# use a seperate set of sqe_user_data range for openat and openat2.
+type sqe_user_data_openat int64[0:3]
+type sqe_user_data_not_openat int64[4:10]
+type sqe_user_data int64[0:10]
+
+# The buffer id (bid) and the buffer group id (bgid) are registered using
+# IORING_OP_PROVIDE_BUFFERS. Use the ids in a limited range to ease collisions
+# with other operations.
+type io_uring_bid[T] T[0:10]
```

Comment:
I would limit them even more. We generally use 0-3 or 0-4 in similar cases.
4-5 buffers should be enough to trigger bug of any complexity. But reducing the range we will guide the fuzzer towards more interesting interactions. Getting non-matching ids will still be trivial for the fuzzer, I does not need lots of help there.

Reply:
[fixed](https://github.com/google/syzkaller/pull/1926/commits/698f056b535e57897347906c27af968a9e5e7543#diff-2e824a90609ef73cc6319dbc406b6b5eR293-R294)